### DISCUSSION TITLE
FastAPI gets stuck when using run_in_threadpool

### DISCUSSION THREAD

**Discussion by iudeen** [fastapi/fastapi#6512] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def data_fetch(query):
    # I take a few seconds to respond
    return 

@app.get('/my_home')
async def fetch(somedata: ReqData):
    await run_in_threadpool(data_fetch, somedata.q1)
    await run_in_threadpool(data_fetch, somedata.q2)
    return True
```

### Description

I have implemented all my routes using async. And have followed all guidelines from the FastAPI documentation. 

Every route has multiple DB calls, which does not have async support, so they are normal function like this

```
def data_fetch(query):
    # I take a few seconds to respond
    return 
```

To avoid blocking my event loop I use `fastapi.concurrancy.run_in_threadpool`

Now the issue is, when a large number of requests come, my new requests are getting blocked. Even if I close the browser tab (cancel request), the entire app gets stuck till the older requests get processed.

What am I doing wrong here?

I use `uvicorn` as my ASGI server. I run in a kubernetes cluster with 2 replica.

_Few suspects: Am I spawning too many threads? Is it some bug within uvicron? Not really sure!_

### Operating System

Linux

### FastAPI Version

0.73

### Python Version

3.9

**Comment by jgould22** [comment]
Perhaps this comment would be helpful to you https://github.com/tiangolo/fastapi/issues/603#issuecomment-540834120

**Comment by iudeen** [comment]
> Perhaps this comment would be helpful to you [#603 (comment)](https://github.com/tiangolo/fastapi/issues/603#issuecomment-540834120)

@jgould22 I saw that, added the work around by setting `max_workers` on start-up. I need to check how it performs under load. Meanwhile, does [this](https://stackoverflow.com/a/70929141/10394971) also resolve the issue?

**Comment by Kludex** [comment]
Can I see a minimal reproducible example on which it gets stuck?

**Comment by iudeen** [comment]
@Kludex sorry for the late reply. Here's a reproducible example. When a lot of request comes, ~30 per minute, the entire process gets stuck till the container is restarted. 

_note: there can be multiple `run_in_threadpool` calls happening to fetch various things from Mongo._

We serve using unvicorn in Kubernetes pod with 2 CPUs. 

```py
import httpx
import pandas as pd
import pymongo
from fastapi import FastAPI
from fastapi.concurrency import run_in_threadpool

app = FastAPI()


async def get_data() -> dict:
    async with httpx.AsyncClient() as client:
        r = await client.get('https://www.example.com/')
        # big json ~20mb, sometimes can take 10-20 seconds for response
        return r.json()


def get_meta_data(doc_id: int) -> dict:
    client = pymongo.MongoClient("mongodb://localhost:27017")
    db = client["test"]
    collection = db["docs"]
    return collection.find_one({"doc_id": doc_id}) # sometimes takes ~4-5 seconds for response 


@app.get("/get_data")
async def get_data(doc_id: int):
    data = get_data()
    df = pd.DataFrame.from_records(data)

    # some df operations using pandas and numpy mostly

    meta = await run_in_threadpool(get_meta_data, doc_id)
    df.rename(columns=meta.get("columns"))
    return df.to_dict(orient='records')
```

**Comment by raphaelauv** [comment]
how async and python work -> https://github.com/tiangolo/fastapi/issues/4265#issuecomment-990945629
